packages/beacon-node/src/metrics/validatorMonitor.ts

```diff
@@ -156,26 +180,47 @@ function withEpochSummary(validator: MonitoredValidator, epoch: Epoch, fn: (summ
     validator.summaries.set(epoch, summary);
   }
 
-  fn(summary);
-
   // Prune
-  const toPrune = validator.summaries.size - HISTORIC_EPOCHS;
+  const toPrune = validator.summaries.size - MAX_CACHED_EPOCHS;
   if (toPrune > 0) {
     let pruned = 0;
     for (const idx of validator.summaries.keys()) {
       validator.summaries.delete(idx);
       if (++pruned >= toPrune) break;
     }
   }
+
+  return summary;
 }
 
+// To uniquely identify an attestation:
+// `index=$validator_index target=$target_epoch:$target_root
+type AttestationSummary = {
+  poolSubmitDelayMinSec: number | null;
+  poolSubmitSentPeers: number | null;
+  aggregateInclusionDelaysSec: number[];
+  blockInclusions: AttestationBlockInclusion[];
+};
+
+type AttestationBlockInclusion = {
+  blockRoot: RootHex;
+  blockSlot: Slot;
+  votedCorrectHeadRoot: boolean;
+  votedForMissedSlot: boolean;
+  attestationSlot: Slot;
+};
+
+/** `$target_epoch:$target_root` */
+type TargetRoot = string;
+
 /// A validator that is being monitored by the `ValidatorMonitor`. */
 type MonitoredValidator = {
-  /// The validator index in the state. */
-  index: number;
   /// A history of the validator over time. */
   summaries: Map<Epoch, EpochSummary>;
   inSyncCommitteeUntilEpoch: number;
+  // Unless the validator slashes itself, there MUST be one attestation per target checkpoint
+  attestations: MapDefMax<Epoch, MapDefMax<TargetRoot, AttestationSummary>>;
```

Comment:
This PR assumes:
- There's strictly one attestation per tuple of (validator index, target_checkpoint)
- If validator submits the same attestation twice, the second event is ignored
- An attestation can be included in multiple blocks, which may or may not be cannonical